### PR TITLE
Advertise driver-specific addresses

### PR DESCRIPTION
**Pull request by schmichael**
# Summary
* DriverNetwork struct returned by Driver.Prestart and Driver.Start containing the PortMap and IP set by the Driver.
* `NOMAD_{HOST,DRIVER}_{ADDR,IP,PORT}_<label>` env vars
* `service.address_mode` to determine what IP to advertise

## DriverNetwork

Ideally this would only need to be returned by Prestart, but Docker doesn't assign a container an IP until the container starts.

I could have implemented it such that Prestart returns PortMap and Start returns IP/AutoAdvertise, but here's why I didn't:

* DriverNetwork is a really nice way to wrap up all of the container network metadata for passing through TaskRunner to Consul ServiceClient. Even if the Driver didn't use it I'd probably keep it around in TaskRunner and Consul.
* In theory other drivers may work differently and provide the IP in Prestart. I couldn't actually find where Docker documented their behavior, so it's conceivable they'll even change someday.
* This mirrors Docker's own Inspect function: it returns networking information whether it's populated or not.

The last point raises an interesting question: **Should I just create `Driver.Inspect` and have it return `DriverNetwork?** TaskRunner would just call it as needed and we wouldn't need to persist DriverNetwork.

Might make for a good followup PR.

## `NOMAD_{HOST,DRIVER}_{ADDR,IP,PORT}_<label>` env vars

This was the Cambrian explosion of networking env vars. My work revealed something interesting: `NOMAD_ADDR_<label>` isn't reliable. If a `port_map` is set it concatenates the **host** IP with the **driver** port. This can only ever be right by pure happenstance.

So I marked it as deprecated and introduced explicit env vars.

The *big* caveat here is that the DRIVER_IP (and therefore ADDR) env vars can't be populated until the driver returns them -- which right now only happens *after Start.* This makes it only really useful for script checks. Everything else happens before Start is called.

In theory if other drivers return an IP from Prestart then it would be available in Start (eg `args` interpolation).

I tried to document this well, but it feels like a pretty big footgun.

## `service.address_mode`

Turned out as we hoped! `auto`, `host`, or `driver` where `auto` means `host` unless you're using a Docker networking plugin.

**Checks always use host IP+Port** as we discussed. A bit of a gotcha, but I think it's always the right choice. If it needs to be configurable for esoteric setups we can add that later.

## Naming (Driver vs Plugin)

*As always I'm willing to changes any names!* That being said I went with *driver* everywhere as it just made sense. It's the network defined by the driver. I think using "plugin" right now would just leave people hunting around for what these "plugins" are -- or conflate it with Docker plugins which is kind of true now but would probably lead to confusion in the future?

### CODE REVIEW

File: client/driver/env/env.go [hunk 1]
```diff
@@ -56,20 +57,50 @@ const (
...
+
 	// HostPortPrefix is the prefix for passing the host port when a portmap is
 	// specified.
 	HostPortPrefix = "NOMAD_HOST_PORT_"
 
+	// DriverAddrPrefix is the prefix for passing the address of port
+	// allocations to tasks with the driver's IP address if on exists.
+	// Service's will advertise this address if address_mode=true or
+	// address_mode=auto and the driver determines it should be used.
+	DriverAddrPrefix = "NOMAD_DRIVER_ADDR_"
```
Comment (dadgar) [inline client/driver/env/env.go]: Remove

File: client/driver/env/env.go [hunk 2]
```diff
@@ -56,20 +57,50 @@ const (
...
 
 	// AddrPrefix is the prefix for passing both dynamic and static port
-	// allocations to tasks.
+	// allocations to tasks
 	// E.g $NOMAD_ADDR_http=127.0.0.1:80
+	//
+	// Deprecated: Use NOMAD_HOST_ADDR_ or NOMAD_DRIVER_ADDR_ instead as
+	// this environment variable will only ever use the host IP. If a port
+	// map is used this variable will be set to the Host IP and Driver's
+	// Port which likely won't work in either context.
```
Comment (dadgar) [inline client/driver/env/env.go]: Make it always be host_ip:host_port and don't say deprecated

File: client/driver/env/env.go [hunk 3]
```diff
@@ -56,20 +57,50 @@ const (
...
 	AddrPrefix = "NOMAD_ADDR_"
 
-	// IpPrefix is the prefix for passing the IP of a port allocation to a task.
+	// IpPrefix is the prefix for passing the IP of a port allocation to a
+	// task. This may not be the host's address depending on task
```
Comment (dadgar) [inline client/driver/env/env.go]: Always host IP

File: client/driver/env/env.go [hunk 4]
```diff
@@ -56,20 +57,50 @@ const (
...
+	// configuration.
 	IpPrefix = "NOMAD_IP_"
 
 	// PortPrefix is the prefix for passing the port allocation to a task.
 	PortPrefix = "NOMAD_PORT_"
 
+	// HostAddrPrefix is the prefix for passing both dynamic and static
+	// port allocations to tasks with the host's IP address for cases where
+	// the task advertises a different address.
```
Comment (dadgar) [inline client/driver/env/env.go]: Remove these

File: client/driver/qemu.go [hunk 5]
```diff
@@ -276,7 +276,7 @@ func (d *QemuDriver) Start(ctx *ExecContext, task *structs.Task) (DriverHandle,
 		waitCh:         make(chan *dstructs.WaitResult, 1),
 	}
 	go h.run()
-	return h, nil
+	return &StartResponse{Handle: h}, nil
```
Comment (schmichael) [inline client/driver/qemu.go]: We've only ever propogated the port map for the Docker driver despite rkt and qemu also having port maps. Should I go ahead and add that as part of this PR? Should Just Work but would be a new feature (or bugfix depending on perspective).

File: api/tasks.go [hunk 6]
```diff
@@ -95,11 +95,12 @@ type ServiceCheck struct {
...
-	Name      string
-	Tags      []string
-	PortLabel string `mapstructure:"port"`
-	Checks    []ServiceCheck
+	Id          string
+	Name        string
+	Tags        []string
+	PortLabel   string `mapstructure:"port"`
+	AddressMode string `mapstructure:"address_mode"`
+	Checks      []ServiceCheck
```
Comment (schmichael) [inline api/tasks.go]: Added

File: client/driver/docker.go [hunk 7]
```diff
@@ -585,7 +600,51 @@ func (d *DockerDriver) Start(ctx *ExecContext, task *structs.Task) (DriverHandle
...
+			continue
+		}
+		n++
+		ip = net.IPAddress
+		ipName = name
+
+		// Don't auto-advertise bridge IPs
+		if name != "bridge" {
+			auto = true
+		}
```
Comment (schmichael) [inline client/driver/docker.go]: Done

File: client/driver/qemu.go (hunk 5 above)
Comment (schmichael) [inline client/driver/qemu.go]: From offline discussion: going to have rkt and qemu return their port maps.

**Comment by schmichael** [comment]
@dadgar Ready for round 2.

File: client/driver/docker.go [hunk 8]
```diff
@@ -585,7 +600,50 @@ func (d *DockerDriver) Start(ctx *ExecContext, task *structs.Task) (DriverHandle
...
+		Network: &cstructs.DriverNetwork{
+			PortMap:       d.driverConfig.PortMap,
+			IP:            ip,
+			AutoAdvertise: autoUse,
+		},
+	}
+	return resp, nil
+}
+
+func (d *DockerDriver) detectIP(c *docker.Container) (string, bool) {
```
Comment (dadgar) [inline client/driver/docker.go]: Put a comment on what this does and the fields it returns

File: client/driver/docker.go [hunk 9]
```diff
@@ -585,7 +600,50 @@ func (d *DockerDriver) Start(ctx *ExecContext, task *structs.Task) (DriverHandle
...
+	if c.NetworkSettings == nil {
+		// This should only happen if there's been a coding error (such
+		// as not calling InspetContainer after CreateContainer). Code
+		// defensively in case the Docker API changes subtly.
+		d.logger.Printf("[ERROR] driver.docker: no network settings for container %s", c.ID)
+		return "", false
+	}
+	ip, ipName := "", ""
```
Comment (dadgar) [inline client/driver/docker.go]: Add blank spaces between blocks. Here and line 643

File: client/driver/env/env.go [hunk 10]
```diff
@@ -455,17 +451,52 @@ func (b *Builder) SetSecretsDir(dir string) *Builder {
...
-	b.portMap = newPortMap
+	b.driverNetwork = ncopy
 	b.mu.Unlock()
 	return b
 }
 
+// buildNetworkEnv env vars in the given map.
+//
+//	Auto:   NOMAD_{IP,PORT,ADDR}_<label>
+//	Host:   NOMAD_HOST_{IP,PORT,ADDR}_<label>
```
Comment (dadgar) [inline client/driver/env/env.go]: Update comment

File: nomad/structs/structs.go [hunk 11]
```diff
@@ -2430,6 +2450,12 @@ func (sc *ServiceCheck) Hash(serviceID string) string {
 	return fmt.Sprintf("%x", h.Sum(nil))
 }
 
+const (
+	AddressModeAuto   = "auto"
```
Comment (dadgar) [inline nomad/structs/structs.go]: Comments

File: website/source/docs/drivers/docker.html.md [hunk 12]
```diff
@@ -401,6 +401,25 @@ Note that by default this only works with `bridged` networking mode. It may
...
+### Using Docker IPs and Ports
+
+*New in Nomad 0.6.*
+
+Other than `host` and `none`, Docker network modes create an IP for the
+container. This IP is set when the container is run and exposed as a
+`NOMAD_DRIVER_IP_<label>` environment variable for use in script checks.
+
+The ports specified in the `port_map` are exposed via the
+`NOMAD_DRIVER_PORT_<label>` environment variables. Unlike the IP, ports are
```
Comment (dadgar) [inline website/source/docs/drivers/docker.html.md]: Update

**Pull request merged**